### DISCUSSION TITLE
Unbuffered stdout

### DISCUSSION THREAD

**Discussion by novakboskov** [esnet/iperf#1337] [Q&A] [answered]
I invoke `iperf3` from a C++ via `popen`:
```
FILE *iperf_pipe = popen("iperf3 -c <SOME_IP> -t 1 | awk 'NR==4 {printf \"%s %s\\n\", $(NF - 4), $(NF - 3); system(\"\")}'", "r");
char *buffer[128];
if (fgets(buffer, 128, iperf_pipe) == NULL)
   exit 1;
court << buffer;
``` 
Note that `-t 1` option.

And the output differs depending on the network speed. When `<SOME_IP>=localhost` then a number is printed (e.g., `12Gbits/s`, because `localhost`). When `iperf3` needs to go over a slow connection (some `Kbps`), then `fgets` does not read anything from `iperf_pipe`, and the program exits.

To make the issue even more interesting, when I run from a tty:
```
$ iperf3 -c <SOME_IP> -t 1 | awk 'NR==4 {printf "%s %s\n", $(NF - 4), $(NF - 3); system("")}'
```
It always prints the network speed, regardless of whether the connection is fast or slow.

This makes me think that `iperf3` buffers its `stdout` depending on whether it goes to a tty. **How do I force `iperf3` to flush its stdout as soon as possible, even when I'm not on a tty?**

I use the following `iperf3` version:
```
iperf 3.1.3
Linux sg 5.4.0-110-generic #124~18.04.1-Ubuntu SMP Fri Apr 22 12:01:21 UTC 2022 x86_64
Optional features available: CPU affinity setting, IPv6 flow label, TCP congestion algorithm setting, sendfile / zerocopy, socket pacing
```
Unfortunately, I'd have to stick to it unless there is a self-contained binary of a newer `iperf3` version.

**Comment by davidBar-On** [comment]
Some `flash` commands were added in later iperf3 versions.  You can with newer iperf3 version for Windows from [here](https://files.budman.pw/) (not official site - maintained by [BudMan](https://www.neowin.net/forum/tags/iperf3/)).

**Reply by novakboskov** [reply]
Thanks!

**Comment by novakboskov** [comment, marked as answer]
What I ended up doing is actually two things:
- installed the latest version of `iperf3`, it's `3.10` (compiling it from source from the official website, Linux build), and
- got rid of `awk`. Now I read `iperf3` output directly via `fgets` and parse the text on my own.

Apparently, the buffering problem is not there anymore.
